File: source/downloads/code/LeetCode/Python/4Sum.py
 class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if not nums:
            return nums

        nums = sorted(nums)
        ret = []
        n = len(nums)
        for i in range(0,n-3):
            # skip all redundant element
            if i!=0 and nums[i] == nums[i-1]:
                continue
            
            for j in range(n-1,i+2,-1):
                # skipp repeated elements
                if j != n-1 and nums[j] == nums[j+1]:
                    continue
                l = i+1
                r = j-1
                while l < r:
                    val = nums[i]+nums[j]+nums[l]+nums[r]
                    if val == target:
                        ret.append([nums[i],nums[l],nums[r],nums[j]])
                        l += 1
                        r -= 1
                        # skip repeated element
                        while l<r and nums[l] == nums[l-1]:
                            l += 1
                        while l<r and nums[r] == nums[r+1]:
                            r = r-1
                            
                    elif val < target:
                        l = l+1
                    else:
                        r = r-1
        return ret      

File: source/downloads/code/LeetCode/Java/4Sum.java
public class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        if(nums.length < 4) return ret;
        Arrays.sort(nums);
        int n = nums.length;
        for(int i=0;i<n-3;i++){
            // skip 重复nums[i]
            if(i>0 && nums[i] == nums[i-1]) continue;
            for(int j=i+1;j<n-2;j++){
               // skip重复的nums[j]
               if(j>i+1 && nums[j] == nums[j-1]) continue;
               int low = j+1;
               int high = n-1;

               // j, low, high　组成3sum
                while (low<high){
                    int sum = nums[i]+nums[j]+nums[low]+nums[high];
                    if(sum==target){
                        ret.add(Arrays.asList(nums[i],nums[j],nums[low],nums[high]));
                        //skip 重复的nums[low] nums[high]
                        while(low<high && nums[low] == nums[low+1]) low++;
                        while(low<high && nums[high] == nums[high-1]) high--;
                        low++;
                        high--;
                    }
                    else if(sum < target) low++;
                    else    high--;
                
                }
            
            
            }
        
        
        
        }
        return ret;
    }
}
